server/src/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1120,14 +1123,60 @@ public void cleanupStorage(boolean recurring) {
                     }
 
                     // remove snapshots in Error state
-                    List<SnapshotVO> snapshots = _snapshotDao.listAllByStatus(Snapshot.State.Error);
+                    List<SnapshotVO> snapshots = _snapshotDao.listAllByStatusIncludingRemoved(Snapshot.State.Error);
                     for (SnapshotVO snapshotVO : snapshots) {
                         try {
                             List<SnapshotDataStoreVO> storeRefs = _snapshotStoreDao.findBySnapshotId(snapshotVO.getId());
+                            boolean isVMware = snapshotVO.getHypervisorType().equals(HypervisorType.VMware);
+                            boolean removeSnapshot = true;
                             for (SnapshotDataStoreVO ref : storeRefs) {
-                                _snapshotStoreDao.expunge(ref.getId());
+                                // Cleanup corresponding items (if any) from secondary storage.
+                                if (isVMware) {
```

Comment:
Good catch!

We would have to get the PR in and test it, without the IF, against Xen and KVM, at least. What do you think?